packages/@blockprotocol/graph/src/type-resolver/entity-type-to-typescript.ts

```diff
@@ -0,0 +1,216 @@
+} from "./entity-type-to-typescript/type-definition-generators.js";
+import { hardcodedBpTypes } from "./hardcoded-bp-types.js";
+import { fetchTypeAsJson, generateTypeNameFromSchema } from "./shared.js";
+
+const bannerComment = (uri: string, depth: number) => `/* eslint-disable */
+/**
+ * This file was automatically generated – do not edit it.
+ * @see ${uri} for the root JSON Schema these types were generated from
+ * Types for link entities and their destination were generated to a depth of ${depth} from the root
+ */`;
+
+/**
+ * Uses json-schema-to-typescript to chase down $refs and create a TypeScript type for the provided schema
+ * Ignores 'links' which are followed manually elsewhere in order to manage depth
+ * @todo
+ * - type name is generated from schema title. Clashes append 1, 2 etc. Patch this to distinguish on URI/$id segments?
+ *   @see https://github.com/bcherny/json-schema-to-typescript/blob/34de194e87cd54f43809efae110732569e0891c1/src/parser.ts#L303
+ *
+ */
+export const compileSchema = (schema: any, options?: Partial<Options>) =>
+  compile(schema, "This value doesn't appear to matter", {
+    strictIndexSignatures: true,
+    // @see https://apitools.dev/json-schema-ref-parser/docs/options.html
+    $refOptions: {
+      dereference: {
+        circular: true,
+        onDereference: () => null, // @todo why is this required
+      },
+      resolve: {
+        http: {
+          read({ url }) {
+            const hardcodedType =
+              hardcodedBpTypes[url as keyof typeof hardcodedBpTypes];
+            if (hardcodedType) {
+              return hardcodedType;
+            }
+            return fetchTypeAsJson(url);
+          },
+        },
+      },
+    },
+    ...options,
+  });
+
+// A named type and the TS string defining it and its type dependencies
+type CompiledType = {
+  typeName: string;
+  typeScriptString: string;
+};
+
+// A map of schema URIs to their TS type name and definition
+type UriToType = { [typeUri: string]: CompiledType };
+
+/**
+ * Generates TypeScript types from a given Entity Type schema
+ * If depth > 0, follows 'links' from the schema to include types for link entities and their possible destinations
+ * @param schema the schema to generate types for
+ * @param depth the depth to which the graph of _entity type_ schemas linked from this schema will be resolved
+ * @param resolvedUrisToType a map of already-resolved schemas to skip type generation for
+ */
+const _jsonSchemaToTypescript = async (
+  schema: EntityType,
+  depth: number,
+  resolvedUrisToType: UriToType,
+): Promise<CompiledType & {}> => {
+  if (resolvedUrisToType[schema.$id]) {
+    return resolvedUrisToType[schema.$id]!;
+  }
+
+  // if the cache is empty, this is the schema the external caller provided. we need to know to add boilerplate up top
+  const rootSchema = Object.keys(resolvedUrisToType).length === 0;
+
+  const typeName = generateTypeNameFromSchema(schema);
```

Comment:
Fair enough, done in https://github.com/blockprotocol/blockprotocol/pull/861/commits/19b584bd377bbcb6ac9dec81e11593544d92c4f7 – using `Vn` if it's a different version of the same version, a counter otherwise
![Screenshot 2023-01-05 at 21 15 50](https://user-images.githubusercontent.com/37743469/210883045-176f1adf-5657-4916-9df9-b3b841a950dc.png)